File: lib/src/main/java/net/mm2d/upnp/IconImpl.java
/*
 * Copyright (c) 2018 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp;

import java.io.IOException;
import java.net.URL;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * Iconの実装。
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介(OHMAE Ryosuke)</a>
 */
class IconImpl implements Icon {
    /**
     * DeviceDescriptionのパース時に使用するビルダー
     *
     * @see DeviceParser#loadDescription(HttpClient, DeviceImpl.Builder)
     */
    static class Builder {
        private String mMimeType;
        private int mHeight;
        private int mWidth;
        private int mDepth;
        private String mUrl;

        /**
         * インスタンス作成
         */
        public Builder() {
        }

        /**
         * MimeTypeの値を登録する。
         *
         * @param mimeType MimeType
         * @return Builder
         */
        @Nonnull
        public Builder setMimeType(@Nonnull final String mimeType) {
            mMimeType = mimeType;
            return this;
        }

        /**
         * Heightの値を登録する
         *
         * @param height Height
         * @return Builder
         */
        @Nonnull
        public Builder setHeight(@Nonnull final String height) {
            try {
                mHeight = Integer.parseInt(height);
            } catch (final NumberFormatException e) {
                mHeight = 0;
            }
            return this;
        }

        /**
         * Widthの値を登録する。
         *
         * @param width Width
         * @return Builder
         */
        @Nonnull
        public Builder setWidth(@Nonnull final String width) {
            try {
                mWidth = Integer.parseInt(width);
            } catch (final NumberFormatException e) {
                mWidth = 0;
            }
            return this;
        }

        /**
         * Depthの値を登録する
         *
         * @param depth Depth
         * @return Builder
         */
        @Nonnull
        public Builder setDepth(@Nonnull final String depth) {
            try {
                mDepth = Integer.parseInt(depth);
            } catch (final NumberFormatException e) {
                mDepth = 0;
            }
            return this;
        }

        /**
         * URLの値を登録する。
         *
         * @param url URL
         * @return Builder
         */
        @Nonnull
        public Builder setUrl(@Nonnull final String url) {
            mUrl = url;
            return this;
        }

        /**
         * Iconのインスタンスを作成する。
         *
         * @return Iconのインスタンス
         * @throws IllegalStateException 必須パラメータが設定されていない場合
         */
        @Nonnull
        public Icon build() throws IllegalStateException {
            if (mMimeType == null) {
                throw new IllegalStateException("mimetype must be set.");
            }
            if (mWidth <= 0) {
                throw new IllegalStateException("width must be > 0.");
            }
            if (mHeight <= 0) {
                throw new IllegalStateException("height must be > 0.");
            }
            if (mDepth <= 0) {
                throw new IllegalStateException("depth must be > 0.");
            }
            if (mUrl == null) {
                throw new IllegalStateException("url must be set.");
            }
            return new IconImpl(this);
        }
    }

    @Nonnull
    private final String mMimeType;
    private final int mHeight;
    private final int mWidth;
    private final int mDepth;
    @Nonnull
    private final String mUrl;
    @Nullable
    private byte[] mBinary;

    /**
     * インスタンスを作成する。
     *
     * @param builder Builder
     */
    private IconImpl(@Nonnull final Builder builder) {
        mMimeType = builder.mMimeType;
        mHeight = builder.mHeight;
        mWidth = builder.mWidth;
        mDepth = builder.mDepth;
        mUrl = builder.mUrl;
    }

    @Override
    @Nonnull
    public String getMimeType() {
        return mMimeType;
    }

    @Override
    public int getHeight() {
        return mHeight;
    }

    @Override
    public int getWidth() {
        return mWidth;
    }

    @Override
    public int getDepth() {
        return mDepth;
    }

    @Override
    @Nonnull
    public String getUrl() {
        return mUrl;
    }

    /**
     * URLからバイナリデータを読み込む。
     *
     * @param device DeviceImpl
     * @param client 通信に使用する{@link HttpClient}
     * @throws IOException 通信エラー
     */
    void loadBinary(
            @Nonnull final DeviceImpl device,
            @Nonnull final HttpClient client) throws IOException {
        final URL url = device.getAbsoluteUrl(mUrl);
        mBinary = client.downloadBinary(url);
    }

    @Override
    @Nullable
    public byte[] getBinary() {
        return mBinary;
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/internal/impl/IconImpl.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.impl

import net.mm2d.upnp.Http
import net.mm2d.upnp.Icon
import net.mm2d.upnp.SingleHttpClient
import java.io.IOException

/**
 * Implements for [Icon].
 *
 * @author [大前良介(OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
internal class IconImpl(
    override val mimeType: String,
    override val height: Int,
    override val width: Int,
    override val depth: Int,
    override val url: String
) : Icon {
    override var binary: ByteArray? = null
        private set

    @Throws(IOException::class)
    fun loadBinary(client: SingleHttpClient, baseUrl: String, scopeId: Int) {
        val url = Http.makeAbsoluteUrl(baseUrl, url, scopeId)
        binary = client.downloadBinary(url)
    }

    internal class Builder {
        private var mimeType: String? = null
        private var height: Int = 0
        private var width: Int = 0
        private var depth: Int = 0
        private var url: String? = null

        @Throws(IllegalStateException::class)
        fun build(): IconImpl {
            val mimeType = mimeType
                ?: throw IllegalStateException("mimetype must be set.")
            check(width > 0) { "width must be > 0. actually $width" }
            check(height > 0) { "height must be > 0. actually $height" }
            check(depth > 0) { "depth must be > 0. actually $depth" }
            val url = url
                ?: throw IllegalStateException("url must be set.")

            return IconImpl(
                mimeType = mimeType,
                height = height,
                width = width,
                depth = depth,
                url = url
            )
        }

        fun setMimeType(mimeType: String): Builder = apply {
            this.mimeType = mimeType
        }

        fun setHeight(height: String): Builder = apply {
            this.height = height.toIntOrNull() ?: 0
        }

        fun setWidth(width: String): Builder = apply {
            this.width = width.toIntOrNull() ?: 0
        }

        fun setDepth(depth: String): Builder = apply {
            this.depth = depth.toIntOrNull() ?: 0
        }

        fun setUrl(url: String): Builder = apply {
            this.url = url
        }
    }
}
